### PR TITLE
JAMES-3359 Mailbox/set parentId updates

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java [hunk 1]
```diff
@@ -145,6 +147,14 @@ public MailboxPath child(String childName, char delimiter) {
         return levels;
     }
 
+    /**
+     * @return the name of the mailbox, accounting for the delimiter
+     */
+    public String getName(char delimiter) {
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java:153]: Needs tests

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala [hunk 2]
```diff
@@ -77,6 +108,21 @@ case class MailboxPatchObject(value: Map[String, JsValue]) {
...
+object ValidatedMailboxPatchObject {
+  val nameProperty: NonEmptyString = "/name"
+  val parentIdProperty: NonEmptyString = "/parentId"
+}
+
+case class ValidatedMailboxPatchObject(nameUpdate: Option[NameUpdate],
+                                       parentIdUpdate: Option[ParentIdUpdate]) {
+  val shouldUpdateMailboxPath: Boolean = nameUpdate.isDefined || parentIdUpdate.isDefined
+
+  val properties: Properties = Properties(List(
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala:120]: updatedProperties?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 3]
```diff
@@ -2911,4 +2911,629 @@ trait MailboxSetMethodContract {
...
+         |        "isSubscribed": false
+         |      }],
+         |      "notFound": []
+         |    }, "c2"]
+         |  ]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def updateShouldAllowParentIdChangeWhenTopLevelMailboxAndNewMane(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:3147]:
```suggestion
  def updateShouldAllowParentIdChangeWhenTopLevelMailboxAndNewName(server: GuiceJamesServer): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 4]
```diff
@@ -2911,4 +2911,629 @@ trait MailboxSetMethodContract {
...
+         |    }, "c2"]
+         |  ]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def updateShouldAllowParentIdDropWhenChildMailbox(server: GuiceJamesServer): Unit = {
+    val parentId: MailboxId = server.getProbe(classOf[MailboxProbeImpl])
+      .createMailbox(MailboxPath.forUser(BOB, "parent"))
+    val newParentId: MailboxId = server.getProbe(classOf[MailboxProbeImpl])
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:3330]: No need?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 5]
```diff
@@ -2911,4 +2911,629 @@ trait MailboxSetMethodContract {
...
+         |        "isSubscribed": false
+         |      }],
+         |      "notFound": []
+         |    }, "c2"]
+         |  ]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def updateShouldAllowParentIdDropWhenTopLevelMailboxAndNewMane(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:3416]:
```suggestion
  def updateShouldAllowParentIdDropWhenTopLevelMailboxAndNewName(server: GuiceJamesServer): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 6]
```diff
@@ -2911,4 +2911,629 @@ trait MailboxSetMethodContract {
...
+         |    }, "c2"]
+         |  ]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def updateShouldAllowParentIdDropWhenChildMailboxAndNewName(server: GuiceJamesServer): Unit = {
+    val parentId: MailboxId = server.getProbe(classOf[MailboxProbeImpl])
+      .createMailbox(MailboxPath.forUser(BOB, "parent"))
+    val newParentId: MailboxId = server.getProbe(classOf[MailboxProbeImpl])
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:3507]: No need?

**Comment by chibenwa** [comment]
```
[d8f612d7e7dd687d0251383996541b0a13fdf30e] [ERROR] Failed to execute goal on project james-json: Could not resolve dependencies for project org.apache.james:james-json:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.fasterxml.jackson.datatype:jackson-datatype-guava:jar:2.9.9 -> com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:animal-sniffer-parent:pom:1.14 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/animal-sniffer-parent/1.14/animal-sniffer-parent-1.14.pom.part (No such file or directory) -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
test this please

` error reading /root/.m2/repository/com/google/guava/guava/25.1-jre/guava-25.1-jre.jar; zip file is empty`

**Comment by chibenwa** [comment]
`RemoteDeliveryErrorHandlingTest.remoteDeliveryShouldStoreTemporaryFailureAsPermanentWhenExceedsMaximumRetries:268 » ConditionTimeout`

Unrelated and we should have a look

test this please

**Comment by LanKhuat** [comment]
There are still some indent mistakes?

**Comment by chibenwa** [comment]
Didn't the last commit solved that?

**Comment by LanKhuat** [comment]
> Didn't the last commit solved that?

Not all of them. Some are still wrong by 1 space

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 7]
```diff
@@ -1777,18 +1777,18 @@ trait MailboxSetMethodContract {
...
-        .log().ifValidationFails()
-        .statusCode(SC_OK)
-        .contentType(JSON)
-        .extract
-        .body
-        .asString
+    val response = `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+     .when
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1783]: Like here

**Comment by chibenwa** [comment]
Should be good now, I did a full file review in my IDE

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
```
[2067aef2db6c7e1c05ad89a94fe6048141e5fc02] [ERROR] error reading /root/.m2/repository/com/github/fge/throwing-lambdas/0.5.0/throwing-lambdas-0.5.0.jar; zip file is empty
[2067aef2db6c7e1c05ad89a94fe6048141e5fc02] [ERROR] /james-project/json/src/main/java/org/apache/james/json/DTOModule.java:[20,1] cannot access org.apache.james.json
[2067aef2db6c7e1c05ad89a94fe6048141e5fc02] [ERROR]   zip END header not found
```

test this please

**Comment by chibenwa** [comment]
`java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:34863`

test this please

**Review by Arsnael — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 8]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+         |                      "/parentId": "${parentId.serialize()}"
+         |                    }
+         |                }
+         |           },
+         |    "c1"
+         |       ],
+         |       ["Mailbox/get",
+         |         {
+         |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5156]: add properties filtering to minimize the response block :)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 9]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+         |                      "/parentId": "${newParentId.serialize()}"
+         |                    }
+         |                }
+         |           },
+         |    "c1"
+         |       ],
+         |       ["Mailbox/get",
+         |         {
+         |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5244]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 10]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+         |                      "/name": "newName"
+         |                    }
+         |                }
+         |           },
+         |    "c1"
+         |       ],
+         |       ["Mailbox/get",
+         |         {
+         |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5332]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 11]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5421]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 12]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+         |                      "/parentId": null
+         |                    }
+         |                }
+         |           },
+         |    "c1"
+         |       ],
+         |       ["Mailbox/get",
+         |         {
+         |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5509]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 13]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+         |                      "/name": "newName"
+         |                    }
+         |                }
+         |           },
+         |    "c1"
+         |       ],
+         |       ["Mailbox/get",
+         |         {
+         |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5596]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 14]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5684]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 15]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+        |                      "/name": "mailbox"
+        |                    }
+        |                }
+        |           },
+        |    "c1"
+        |       ],
+        |       ["Mailbox/get",
+        |         {
+        |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5769]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 16]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+        |                      "/parentId": "${parentId.serialize()}"
+        |                    }
+        |                }
+        |           },
+        |    "c1"
+        |       ],
+        |       ["Mailbox/get",
+        |         {
+        |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5856]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 17]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+        |                      "/parentId": "#C42"
+        |                    }
+        |                }
+        |           },
+        |           "c2"
+        |       ],
+        |       ["Mailbox/get",
+        |         {
+        |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |           "ids": ["${mailboxId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:5953]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 18]
```diff
@@ -5073,4 +5126,1547 @@ trait MailboxSetMethodContract {
...
+        |                      "/parentId": "${parentId.serialize()}"
+        |                    }
+        |                }
+        |           },
+        |           "c1"
+        |       ],
+        |       ["Mailbox/get",
+        |         {
+        |           "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |           "ids": ["${parentId.serialize()}"]
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:6549]: idem

**Review by nvduc91 — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala [hunk 19]
```diff
@@ -26,7 +26,11 @@ import eu.timepit.refined.boolean.And
 import eu.timepit.refined.collection.NonEmpty
 import eu.timepit.refined.refineV
 import eu.timepit.refined.string.StartsWith
+<<<<<<< HEAD
 import org.apache.james.core.Username
+=======
+import eu.timepit.refined.types.string.NonEmptyString
+>>>>>>> JAMES-3359 Mailbox/set update should handle existing/not found mailboxes
```
Comment (rouazana) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala]: please fix this commit

**Review by remk — APPROVED** [review]

**Comment by chibenwa** [comment]
Conflict solved.

**Comment by Arsnael** [comment]
Rebased on master and squashed as there was too many conflicts

Added a little extra commit to fix some failing distributed tests as well (the classic mailbox id serialization missing in our jsons)

**Comment by chibenwa** [comment]
test this please (deployment tests)

**Pull request merged**